File: 프로그래머스/lv2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(int m, int n, vector<string> board) {
    int R = board.size(), C = board[0].size();
    int count = 0;
    while (true) {
        int tmp = count;
        bool check[R][C];
        memset(check, 0, sizeof(check));
        for (int i = 0; i < R - 1; i++) {
            for (int j = 0; j < C - 1; j++) {
                if (board[i][j] != ' '
                   && board[i][j] == board[i+1][j]
                   && board[i][j] == board[i][j+1]
                   && board[i][j] == board[i+1][j+1]) {
                    for (int k = 0; k < 4; k++) {
                        check[i][j] = true;
                        check[i+1][j] = true;
                        check[i][j+1] = true;
                        check[i+1][j+1] = true;
                    }
                }
            }
        }
        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                if (check[i][j]) {
                    count++;
                    board[i][j] = ' ';
                }
            }
        }
        if (tmp == count) break;
        for (int i = R - 1; i >= 0; i--) {
            for (int j = 0; j < C; j++) {
                if (board[i][j] == ' ') {
                    for (int idx = i - 1; idx >= 0; idx--) {
                        if (board[idx][j] != ' ') {
                            board[i][j] = board[idx][j];
                            board[idx][j] = ' ';
                            break;
                        }
                    }
                }
            }
        }
    }
    return count;
}

File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.java
import java.util.*;

class Solution {
    public int solution(int m, int n, String[] board) {
        int res = 0;
        char[][] arr = Arrays.stream(board)
            .map(String::toCharArray)
            .toArray(char[][]::new);
        while (true) {
            boolean[][] check = new boolean[m][n];
            for (int i = 0; i < m-1; i++) {
                for (int j = 0; j < n-1; j++) {
                    if (arr[i][j] != ' ' &&
                        arr[i][j] == arr[i+1][j] && 
                        arr[i][j] == arr[i][j+1] && 
                        arr[i][j] == arr[i+1][j+1]) {
                        check[i][j] = true;
                        check[i+1][j] = true;
                        check[i][j+1] = true;
                        check[i+1][j+1] = true;
                    }
                }
            }
            int cnt = 0;
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    if (check[i][j]) {
                        cnt++;
                        arr[i][j] = ' ';
                    }
                }
            }
            if (cnt == 0) break;
            res += cnt;
            for (int i = m - 2; i >= 0; i--) {
                for (int j = 0; j < n; j++) {
                    if (arr[i][j] != ' ' && arr[i+1][j] == ' ') {
                        int idx = i+1;
                        while (idx < m && arr[idx][j] == ' ') idx++;
                        idx--;
                        arr[idx][j] = arr[i][j];
                        arr[i][j] = ' ';
                    }
                }
            }
        }
        return res;
    }
}